newrelic-agent/src/main/java/com/newrelic/agent/stats/dimensional/MeterService.java

```diff
@@ -0,0 +1,314 @@
+
+    @Override
+    public Counter newCounter(String name) {
+        final LongCounter longCounter = meter.get().counterBuilder(name).build();
+        return (increment, attributes) -> longCounter.add(increment, toAttributes(attributes));
+    }
+
+    @Override
+    public Summary newSummary(String name) {
+        // we want a summary, so we call setExplicitBucketBoundariesAdvice with an empty list
+        final DoubleHistogram doubleHistogram = meter.get().histogramBuilder(name)
+                .setExplicitBucketBoundariesAdvice(Collections.emptyList()).build();
+        return (value, attributes) -> doubleHistogram.record(value, toAttributes(attributes));
+    }
+
+    //********************** End Meter APIs *************************/
+
+    static Attributes toAttributes(Map<String,?> map) {
+        final AttributesBuilder builder = Attributes.builder();
+        map.forEach((key, value) -> {
+            if (value instanceof String) {
+                builder.put(key, value.toString());
+            } else if (value instanceof Number) {
+                final Number number = (Number) value;
+                if (value instanceof Float | value instanceof Double) {
+                    builder.put(key, number.doubleValue());
+                } else {
+                    builder.put(key, number.longValue());
+                }
+            } else if (value instanceof Boolean) {
+                builder.put(key, ((Boolean) value).booleanValue());
+            } else {
+                builder.put(key, value.toString());
+            }
+        });
+        return builder.build();
+    }
+
+    private static Map<String, Object> toMap(Attributes attributes) {
+        return attributes.asMap().entrySet().stream().collect(
+                Collectors.toMap(entry -> entry.getKey().getKey(), Map.Entry::getValue));
+    }
+
+    @Override
+    public void harvestEvents(String appName) {
+        try {
+            Collection<MetricData> metricData = metricDataSupplier.get();
+            synchronized (deferredMetricData) {
+                if (!deferredMetricData.isEmpty()) {
+                    metricData = new ArrayList<>(metricData);
+                    metricData.addAll(deferredMetricData);
+                    deferredMetricData.clear();
+                }
+            }
+            sendMetricData(metricData);
+        } catch (Throwable t) {
+            logger.log(Level.FINE, t, "Dimensional metric harvest failed");
+        }
+    }
+
+    /**
+     * Returns true if the send should be retried later.
+     */
+    private void sendMetricData(final Collection<MetricData> metricData) {
+        if (metricData.isEmpty()) {
+            return;
+        }
+        final List<CustomInsightsEvent> events = toEvents(metricData);
+
+
+        logger.log(Level.FINE, "Harvesting {0} dimensional metric events", events.size());
+        try {
+            final int reservoirSize = Math.max(maxSamplesStored, events.size());
```

Comment:
What's going on here? `maxSampledStored` is settable, but doesn't seem to have any impact on behavior outside of being used here. What sets this and what role does the `reservoirSize` play in `RPMService#sendDImensionalMetricData`?